src/System.Management.Automation/engine/parser/tokenizer.cs

```diff
@@ -1229,23 +1233,126 @@ private int SkipBlockComment(int i)
+                case 'r': return "\r";
+                case 't': return "\t";
+                case 'u': return ScanUnicodeEscapeSequence();
+                case 'v': return "\v";
+                default: return c.ToString();
+            }
+        }
+
+        private string ScanUnicodeEscapeSequence()
+        {
+            int escSeqStartIndex = _currentIndex - 2;
+            int maxNumberOfHexDigits = 6;
+            bool isTerminated = false;
+
+            char c = GetChar();
+            if (c != '{')
+            {
+                UngetChar();
+
+                IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                return s_unknownUnicodeChar;
+            }
+
+            // Scan hex chars after the Unicode escape sequence start.
+            var sb = GetStringBuilder();
+            int i;
+            for (i = 0; i < maxNumberOfHexDigits; i++)
+            {
+                c = GetChar();
+
+                // Sequence has been terminated.
+                if (c == '}')
+                {
+                    if (i == 0)
+                    {
+                        // Sequence must have at least one hex char.
+                        IScriptExtent errorExtent = NewScriptExtent(escSeqStartIndex, _currentIndex);
+                        ReportError(errorExtent, () => ParserStrings.InvalidUnicodeEscapeSequence);
+                        return s_unknownUnicodeChar;
+                    }
+
+                    isTerminated = true;
+                    break;
+                }
+
+                if (!c.IsHexDigit())
+                {
+                    UngetChar();
+
+                    ReportError(_currentIndex, () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+
+                sb.Append(c);
+            }
+
+            if (!isTerminated)
+            {
+                c = GetChar();
+                if (c != '}')
+                {
+                    UngetChar();
+                    ReportError(_currentIndex,
+                        i == maxNumberOfHexDigits && c.IsHexDigit()
+                            ? (Expression<Func<string>>)(() => ParserStrings.TooManyDigitsInUnicodeEscapeSequence)
+                            : () => ParserStrings.MissingUnicodeEscapeSequenceTerminator);
+                    return s_unknownUnicodeChar;
+                }
+            }
+
+            string hexStr = GetStringAndRelease(sb);
+
+            int unicodeValue = int.Parse(hexStr, NumberStyles.AllowHexSpecifier, NumberFormatInfo.InvariantInfo);
+            if (unicodeValue <= Char.MaxValue)
+            {
+                return ((char)unicodeValue).ToString();
+            }
+            else if (unicodeValue <= 0x10FFFF)
+            {
+                return Char.ConvertFromUtf32(unicodeValue);
```

Comment:
You shouldn't blindly apply patterns - but it's good to understand when to apply them.

The tokenizer is using the `TryParse` pattern to determine the type of a sequence of characters without raising exceptions. Consider the value `2147483648` - it is convenient to assume it is an `int` because most literals are of type `int`, but when `int.TryParse` fails, we try `long.TryParse` and it succeeds.

When converting hex literals to the numeric value, there is no guessing - the number of digits is sufficient to know type, so we use a different method knowing it can't fail.